[android][jni] Support Tensor MemoryFormat in java wrappers

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40785 [android][jni] Support Tensor MemoryFormat in java wrappers**

The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

IvanKobzarev committed:
[android][jni] Support Tensor MemmoryFormat in java wrappers

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemmoryFormat in java wrappers"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemmoryFormat in java wrappers"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemmoryFormat in java wrappers"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Known problem:
`torch::from_blob` creates first empty Tensor with {0} dimension and if TensorOptions has ChannelsLast - fails.
To create ChannelsLast Tensor from blob I added a separate path:
```
    at::Tensor t = torch::empty(
        torch::IntArrayRef(shapeVec),
        at::TensorOptions(typeMeta).memory_format(
            at::MemoryFormat::ChannelsLast));
    t.unsafeGetTensorImpl()->ShareExternalPointer(...)
```

But during torch::empty(shape) new unused memory is allocated (redundant memory that will not be used, as java buffer memory is used after ShareExternalPointer)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

[ghstack-poisoned]

IvanKobzarev committed:
rebase on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

dreiss commented:
Sorry for the very slow review.  I appreciate the clear explanation of the core change in the commit message!

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][jni] Support Tensor MemoryFormat in java wrappers"


The main goal of this change is to support creating Tensors specifying blob in NHWC (ChannelsLast) format.

ChannelsLast is supported only for 4-dim tensors, this is enforced on LibTorch side, I have not added asserts on java side in case that this limitation will be changed in future and not to have double asserts.

Additional changes in `aten/src/ATen/templates/Functions.h`:

`from_blob` creates `at::empty({0}, options)` tensor first and sets it Storage with sizes and strides afterwards.

But as ChannelsLast is only for 4-dim tensors - it fails on that creation, as dim==1.

I've added `zero_sizes()` function that returns `{0, 0, 0, 0}` for ChannelsLast and ChannelsLast3d.

Differential Revision: [D22396244](https://our.internmc.facebook.com/intern/diff/D22396244)

[ghstack-poisoned]
